### PR TITLE
Memory map feature (first iteration), regions colored by permissions.

### PR DESCRIPTION
**Pull request by sophbohr22**
MemoryMapScreenshot

### CODE REVIEW

File: smaps/src/main/java/com/google/smaps/MemoryMap.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ */
+@WebServlet(name = "MemoryMap", value = "/memorymap")
+public class MemoryMap extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Response will be a Json.
+    response.setContentType("application/json");
+
+    // Parse the file from the user upload.
+    List<Region> regionList = FileParser.getRegionList("/tmp/smaps-upload.txt");
```
Comment (maggieroemer) [inline smaps/src/main/java/com/google/smaps/MemoryMap.java:40]: Hmm I wonder if we can do this just once on upload, and not for each visualization. Maybe leave a todo to investigate?

File: smaps/src/main/webapp/js/memory-map.js [hunk 2]
```diff
@@ -12,4 +12,118 @@
...
+          } else if (perms == 'r--s') {
+            reg.fillStyle = '#FCAD70';  // Orange, 300 intensity.
+          } else if (perms == 'r--p') {
+            reg.fillStyle = '#C58AF9';  // Purple, 300 intensity.
+          } else if (perms == 'rw-s') {
+            reg.fillStyle = '#FF8BCB';  // Pink, 300 intensity.
+          } else if (perms == 'r-xs') {
+            reg.fillStyle = '#669DF6';  // Blue, 400 intensity.
+          }
+          reg.fill();
```
Comment (maggieroemer) [inline smaps/src/main/webapp/js/memory-map.js]: This seems like a good section to pull out into a helper method, maybe "fillColor()"?

**Review by maggieroemer — CHANGES_REQUESTED** [review]
Do you have a memorymap test?

File: smaps/src/main/java/com/google/smaps/MemoryMap.java (hunk 1 above)
Comment (sophbohr22) [reply smaps/src/main/java/com/google/smaps/MemoryMap.java:40]: That's a really good idea, yes I'll look into that.

File: smaps/src/main/webapp/js/memory-map.js (hunk 2 above)
Comment (sophbohr22) [reply smaps/src/main/webapp/js/memory-map.js]: On it!

**Comment by sophbohr22** [comment]
I didn't add a test file yet, I can make one now and add it to this PR.

**Review by maggieroemer — APPROVED** [review]
LGTM, very excited for this feature. Keep it up Sophie!

File: smaps/src/main/java/com/google/smaps/MemoryMap.java (hunk 1 above)
Comment (segfault-the-real-segfault) [reply smaps/src/main/java/com/google/smaps/MemoryMap.java:40]: There must be some kind of AppEngine instanced file system going on if '/tmp/smaps-upload.txt' works every time. It's not a bad thing...just curious and kinda cool.

File: smaps/src/main/java/com/google/smaps/MemoryMap.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    response.getWriter().println(memoryMapJson);
+  }
+
+  /** Creates list of 2D Object arrays of data for memory map. */
+  static ArrayList<Object[]> makeDataArray(List<Region> regions) {
+    // Holds all the arrays of information for the memory map.
+    ArrayList<Object[]> dataArray = new ArrayList<Object[]>();
+
+    // Go through the regions and add range/permissions pairs to the list.
+    for (int i = 0; i < regions.size(); i++) {
```
Comment (segfault-the-real-segfault) [inline smaps/src/main/java/com/google/smaps/MemoryMap.java:61]: At some point, let's add some defenses to the parser so files with extremely large region counts are thrown away. (Say, >10M)

**Review by segfault-the-real-segfault — APPROVED** [review]

**Pull request merged**